File: go/cmd/consume/consumer.go
package main

import (
	"bytes"
	"compress/gzip"
	"crypto/tls"
	"fmt"
	"io"

	"log"
	"net/http"
	"time"

	"github.com/alexflint/go-arg"
	"github.com/segmentio/encoding/json"
	"gopkg.in/ini.v1"

	amqp "github.com/rabbitmq/amqp091-go"
)

func failOnError(err error, msg string) {
	if err != nil {
		log.Panicf("%s: %s", msg, err)
	}
}

// Command line parameters
type CmdArgs struct {
	Config string `arg:"required,-c,--config" help:"The configuration file"`
}

// Configuration file parameters
type RabbitMqConf struct {
	User     string `ini:"user"`
	Password string `ini:"password"`
	Server   string `ini:"server"`
	AmqpPort int    `ini:"amqp_port"`
	HttpPort int    `ini:"http_port"`
	Vhost    string `ini:"vhost"`
	Exchange string `ini:"exchange"`
}

type SplunkConf struct {
	Url   string `ini:"url"`
	Token string `ini:"token"`
}

type Config struct {
	Rabbit  RabbitMqConf
	Splunk  SplunkConf
	Threads int `ini:"threads"`
}

type ProcessResult struct {
	sessionCount int64
}

func GetConfig() *Config {
	// Parse command line arguments to find configuration file
	cmdArgs := &CmdArgs{}
	arg.MustParse(cmdArgs)

	// Now load the configuration file and populate the structures
	config := &Config{}
	iniCfg, err := ini.Load(cmdArgs.Config)
	//cfg, err := ini.Load(cmdArgs.Config)
	failOnError(err, "Failed to parse config file: "+cmdArgs.Config)
	err = iniCfg.Section("general").StrictMapTo(config)
	failOnError(err, "Failed to parse common params: "+cmdArgs.Config)
	err = iniCfg.Section("rabbitmq").StrictMapTo(&config.Rabbit)
	failOnError(err, "Failed to parse rabbitmq params: "+cmdArgs.Config)
	err = iniCfg.Section("splunk").StrictMapTo(&config.Splunk)
	failOnError(err, "Failed to parse splunk params: "+cmdArgs.Config)
	//fmt.Printf("Config: %v\n", config)

	return config
}

type RabbitMq struct {
	conn    *amqp.Connection
	ch      *amqp.Channel
	msgChan <-chan amqp.Delivery
}

func AddVhost(config *RabbitMqConf) error {

	// curl -i -u guest:guest -H "content-type:application/json" -XPUT http://localhost:15672/api/vhosts/dx

	url := fmt.Sprintf("http://%v:%v/api/vhosts/%v", config.Server, config.HttpPort, config.Vhost)
	log.Printf("Adding vhost: %s", url)
	req, err := http.NewRequest("PUT", url, nil)
	if err != nil {
		return err
	}

	req.SetBasicAuth(config.User, config.Password)
	req.Header.Add("content-type", "applicastion/json")

	client := &http.Client{Timeout: 2 * time.Second}
	response, err := client.Do(req)
	if err != nil {
		return err
	}
	defer response.Body.Close()
	// 201 = Created - happens when created
	// 204 = Success No Data - happens if already created
	if response.StatusCode != 201 && response.StatusCode != 204 {
		return fmt.Errorf("Error setting rabbitmq vhost: %v", response.Status)
	}
	return nil
}

func NewRabbitMq(config *RabbitMqConf, threads int) *RabbitMq {
	// Create a new virtual host, so we can login
	err := AddVhost(config)
	failOnError(err, "Failed to create Rabbit MQ vhost")

	connStr := fmt.Sprintf("amqp://%v:%v@%v:%v/%v",
		config.User, config.Password, config.Server, config.AmqpPort, config.Vhost)
	log.Printf(fmt.Sprintf("Connecting to RabbitMQ: %v", connStr))
	conn, err := amqp.Dial(connStr)
	failOnError(err, "Failed to connect to RabbitMQ")

	ch, err := conn.Channel()
	failOnError(err, "Failed to open a channel")

	// Limit the number of unacknowledged messages that can be sent to the consumer.
	// If the prefetch count isn't set, RabbitMQ will continue to send
	// data to the RabbitMQ consumer library, which will buffer it. That will
	// cause the consumer application to continue to consume memory if it can't
	// keep up. It's better to let the messages stay queued in RabbitMQ, where
	// it will be obvious on the RabbitMQ GUI that the consumer isn't keeping
	// up.
	ch.Qos(
		threads*2, // prefetch count
		0,         // prefetch size
		false,     // global
	)

	exchange := config.Exchange
	queue := exchange + "_q"
	// Create the exchange to receive messages
	err = ch.ExchangeDeclare(
		exchange, // name
		"direct", // kind
		true,     // durable
		false,    // auto delete
		false,    // internal
		false,    // no-wait
		nil,      // arguments
	)
	failOnError(err, "Failed to create exchange")

	// Create the Queue to publish messages
	q, err := ch.QueueDeclare(
		queue, // name
		false, // durable
		false, // delete when unused
		false, // exclusive
		false, // no-wait
		nil,   // arguments
	)
	failOnError(err, "Failed to declare a queue")

	// Create a binding to connect the exchange to the Queue
	err = ch.QueueBind(
		queue,    // name
		"",       // key
		exchange, // exchange
		false,    // nowait
		nil,      // arguments
	)
	failOnError(err, "Failed to bind to exchange")

	msgChan, err := ch.Consume(
		q.Name, // queue
		"",     // consumer
		false,  // auto-ack
		false,  // exclusive
		false,  // no-local
		false,  // no-wait
		nil,    // args
	)
	failOnError(err, "Failed to register a consumer")

	log.Printf(fmt.Sprintf("Connected to %v, exchange = %v, queue = %v\n", connStr, exchange, queue))
	return &RabbitMq{conn: conn, ch: ch, msgChan: msgChan}
}

func (p *RabbitMq) MsgChan() <-chan amqp.Delivery {
	return p.msgChan
}

func (p *RabbitMq) Close() error {

	err := p.ch.Close()
	if e2 := p.conn.Close(); e2 != nil && err == nil {
		err = e2
	}

	return err
}

type Splunk struct {
	url    string
	token  string
	client *http.Client
}

func NewSplunk(config *SplunkConf, maxConnsPerHost int) *Splunk {
	tr := &http.Transport{
		MaxIdleConns:        10,
		MaxIdleConnsPerHost: maxConnsPerHost,
		MaxConnsPerHost:     maxConnsPerHost,
		IdleConnTimeout:     30 * time.Second,
		DisableCompression:  false,
		TLSClientConfig:     &tls.Config{InsecureSkipVerify: true}, // Insecure because we don't have certs correct for our Splunk VM
	}
	client := &http.Client{Transport: tr, Timeout: 10 * time.Second}

	return &Splunk{url: config.Url, token: config.Token, client: client}
}

func (p *Splunk) addEvent(builder *bytes.Buffer, json_data map[string]interface{}) error {
	// See https://docs.splunk.com/Documentation/Splunk/8.2.6/Data/HECExamples
	// Splunk requires event data to be wrapped in an "event" element.  It also
	// allows multiple events in a single message. When receiving metadata from
	// Security Analytics, the events will be wrapped in a JSON list. Each item
	// in the list needs to be wrapped in an "event" tag before being sent to
	// splunk.
	// Check if the JSON was an single object or a list
	event := make(map[string]interface{})
	event["event"] = json_data
	// Convert to JSON string
	marshalled, err := json.Marshal(event)
	if err != nil {
		return err
	}

	// Add to final Splunk message
	builder.Write(marshalled)
	return nil
}

func (p *Splunk) convert(json_bytes []byte) ([]byte, int64, error) {
	// Parse the JSON to figure out if it's a list or single object
	var parsedJson interface{}
	sessionCount := int64(0)
	err := json.Unmarshal(json_bytes, &parsedJson)
	if err != nil {
		log.Printf("Failed to Unmarshal: %s", string(json_bytes))
		return nil, sessionCount, err
	}
	builder := bytes.Buffer{}
	switch v := parsedJson.(type) {
	case map[string]interface{}: // JSON object
		err := p.addEvent(&builder, v)
		sessionCount = 1
		if err != nil {
			return nil, sessionCount, err
		}
	case []interface{}: // JSON array
		// Loop through array and add event to each item
		sessionCount = int64(len(v))
		for _, elem := range v {
			switch v := elem.(type) {
			// Only a nested JSON object is valid
			case map[string]interface{}: // JSON object
				err := p.addEvent(&builder, v)
				if err != nil {
					return nil, sessionCount, err
				}
			default:
				return nil, sessionCount, fmt.Errorf("Unexpected nested interface type: %T", elem)
			}
		}
	default:
		return nil, sessionCount, fmt.Errorf("Unexpected nested interface type: %T", parsedJson)
	}
	return builder.Bytes(), sessionCount, nil
}

func (p *Splunk) Send(json_bytes []byte) (int64, error) {

	// Convert the json data to a format that Splunk wants
	convertedBytes, sessionCount, err := p.convert(json_bytes)
	//return session_count, err

	if err != nil {
		return sessionCount, err
	}
	req, err := http.NewRequest("POST", p.url, bytes.NewReader(convertedBytes))
	if err != nil {
		return sessionCount, err
	}

	req.Header.Add("Authorization", "Splunk "+p.token)
	response, err := p.client.Do(req)
	if err != nil {
		return sessionCount, err
	}
	defer response.Body.Close()
	// Always consume the response to enable connection sharing
	b, err := io.ReadAll(response.Body)
	if err == nil && response.Status != "200 OK" {
		responseStr := string(b)
		log.Printf("Splunk Response: %s", responseStr)
	}
	return sessionCount, nil
}

func ProcessMessages(rabbit *RabbitMq, splunk *Splunk, resultChan chan ProcessResult) error {
	sessionCount := int64(0)
	var err error = nil
	for delivery := range rabbit.MsgChan() {
		//log.Printf("Received a message. Content type: %s", d.Body)
		if delivery.ContentEncoding == "gzip" {
			// Decompress the body
			reader := bytes.NewReader(delivery.Body)
			gzreader, err := gzip.NewReader(reader)
			if err != nil {
				log.Printf("Error creating gzip reader: %v", err)
			} else {
				output, err := io.ReadAll(gzreader)
				if err != nil {
					return fmt.Errorf("Error decoding gzip content: %w", err)
				} else {
					//fmt.Println(string(output))
					sessionCount, err = splunk.Send(output)
					if err != nil {
						return fmt.Errorf("Error sending data to Splunk: %w", err)
					}
				}
			}
		} else {
			// Data isn't compressed, so send directly
			sessionCount, err = splunk.Send(delivery.Body)
			if err != nil {
				return fmt.Errorf("Error sending data to Splunk: %w", err)
			}
		}

		delivery.Ack(false)
		// Send back the results to the main thread
		resultChan <- ProcessResult{sessionCount}
	}

	return nil
}

func ForwardData(rabbit *RabbitMq, splunk *Splunk, threads int) {
	totalMsgCount := int64(0)
	lastTotalMsgCount := int64(0)
	totalSessionCount := int64(0)
	lastSessionCount := int64(0)
	lastTime := time.Now().Unix()
	// Create a semaphore channel to limit goroutine concurrency
	resultChan := make(chan ProcessResult)
	// Populate channel with zero so that blocking only occurs after all the threads are running
	//log.Printf("Received a message. Content type: %s, Content Encoding: %s", d.ContentType, d.ContentEncoding)
	// fmt.Print("G")
	for i := 0; i < threads; i++ {
		go ProcessMessages(rabbit, splunk, resultChan)
	}
	// Use up one of the semaphores to process the message
	log.Printf(" [*] Waiting for messages. To exit press CTRL+C")
	for result := range resultChan {
		totalSessionCount += result.sessionCount
		if result.sessionCount != 100 {
			log.Printf("%v Sessions", result.sessionCount)
		}
		totalMsgCount += 1
		now := time.Now().Unix()
		deltaTime := now - lastTime

		if deltaTime >= 5 {
			msgPerSecond := (totalMsgCount - lastTotalMsgCount) / deltaTime
			sessionsPerSecond := (totalSessionCount - lastSessionCount) / deltaTime
			fmt.Println()
			log.Printf("[go] Total Message: %v Messages/Second: %v, Sessions/Second: %v",
				totalMsgCount, msgPerSecond, sessionsPerSecond)
			lastTime = now
			lastTotalMsgCount = totalMsgCount
			lastSessionCount = totalSessionCount
		}
	}
}

func main() {
	// Get the configuration values
	config := GetConfig()
	// Create connection to Rabbit MQ
	rabbit := NewRabbitMq(&config.Rabbit, config.Threads)
	defer rabbit.Close()
	// Create connection to Splunk
	splunk := NewSplunk(&config.Splunk, config.Threads)

	// Start forwarding data from Rabbit MQ to Splunk
	ForwardData(rabbit, splunk, config.Threads)
}

File: python/consumer.py
#!/usr/bin/env python
'''
consumer.py

This script:
    1. Configures a RabbitMQ service to accept Security Analytics
        ICDx alert and metadata messages.
    2. Registers with RabbitMQ as a consumer.
    3. Runs indefinitely, forwarding all RabbitMQ messages
        to a Splunk service.

USAGE
    python3 consumer.py --config [file]

    Configure user arguments in alerts.ini and meta.ini.
'''

import gzip
import sys
import argparse
import configparser
import json
import time
import multiprocessing as mp
import requests
from requests.auth import HTTPBasicAuth
from requests.adapters import HTTPAdapter
import urllib3
import pika

# Debug: uncomment the line below to support HTTP logging
#   and uncomment add_stderr_logger in the new_process function
#from urllib3 import add_stderr_logger


class RabbitToSplunk:
    '''The forwarder process - instantiated by main().'''

    def __init__(self, config, process_num, result_queue):
        '''Set the .ini configuration variables.'''

        self.config = config
        self.process_num = process_num
        self.result_queue = result_queue
        self.msg_count = 0
        self.last_time = int(time.time())
        self.verbose = config['verbose']
        self.debug = config['debug']
        self.session = requests.Session()
        self.session.mount('https://', HTTPAdapter())
        self.session.mount('http://', HTTPAdapter())
        print(f'Forwarding messages from {config["rabbitmq.exchange"]}.'
            f' Process_num = {process_num}')

    def start(self):
        '''
        Configure the RabbitMQ virtual host, exchange, and queues,
            and wait for callbacks. These configuration operations are
            idempotent and may run multiple times during multiple
            consumer launches without any side effects.
        '''

        if self.debug:
            print('\n*** Debug mode is enabled ***')
            time.sleep(3)

        # amqp://<user>:<password>@<server>:<port>/<vhost>
        connection_str = (
                        f'amqp://'
                        f'{self.config["rabbitmq.user"]}:'
                        f'{self.config["rabbitmq.password"]}@'
                        f'{self.config["rabbitmq.server"]}:'
                        f'{self.config["rabbitmq.amqp_port"]}/'
                        f'{self.config["rabbitmq.vhost"]}'
        )
        try:
            connection = pika.BlockingConnection(pika.URLParameters(connection_str))
            channel = connection.channel()
        except ConnectionRefusedError:
            print(f'\n*** The RabbitMQ connection was refused: ({connection_str})...aborting.\n')
            sys.exit(1)

        exchange = self.config['rabbitmq.exchange']

        # Create the exchange to receive messages. This name must match the value in the SA ICDx UI.
        channel.exchange_declare(exchange=exchange, durable=True)

        # Create the Queue to publish messages. The queue name doesn't really matter
        # since this script will connect the queue's exchange, so derive the name
        # from the exchange name by appending "_q".
        queue = exchange + "_q"
        channel.queue_declare(queue=queue)

        # Create a binding to connect the exchange to the Queue
        channel.queue_bind(queue = queue, exchange = exchange, routing_key="")

        # Limit the number of unacknowledged messages that can be sent to the consumer.
        # If the prefetch count isn't set, RabbitMQ will continue to send
        # data to the RabbitMQ consumer library, which will buffer it. That will
        # cause the consumer application to continue to consume memory if it can't
        # keep up. It's better to let the messages stay queued in RabbitMQ, where
        # it will be obvious on the RabbitMQ GUI that the consumer isn't keeping
        # up.
        channel.basic_qos(prefetch_count=10)

        # Start consuming messages
        channel.basic_consume(queue=queue, on_message_callback=self.rabbit_callback, auto_ack=False)

        channel.start_consuming()


    def add_event(self, json_data):
        '''
        Convert each element to a JSON string and add the required "event" tags
            for the HEC data (time, host, source, event).
        '''

        event = {'event': json_data}
        if 'device_time' in json_data.keys():
            # Time must be in seconds.milliseconds. Device time has milliseconds.microseconds
            event['time'] = int(json_data['device_time'])/1000
        if 'device_name' in json_data.keys():
            event['host'] = json_data['device_name']
        if 'product_name' in json_data.keys():
            event['source'] = json_data['product_name']

        event_str = json.dumps(event)

        return event_str


    def send_to_splunk(self, content_encoding, body):
        '''
        Post the message to Splunk

        Splunk requires event data to be wrapped in an "event" element.
        It also allows multiple events in a single message. When receiving
        metadata from Security Analytics, the events will be wrapped in
        a JSON list. Each item in the list needs to be wrapped in
        an "event" tag before being sent to Splunk.

        See https://docs.splunk.com/Documentation/Splunk/8.2.6/Data/HECExamples
        '''

        # Data from Security Analytics is gzipped, so it needs to be decompressed
        if content_encoding == 'gzip':
            # Unzip the data
            json_str = gzip.decompress(body)
            json_str = json_str.decode('utf-8')
        else:
            # Data isn't compressed
            json_str = body.decode('utf-8')


        if self.debug:
            print(f'\nSending data: {json_str}')

        json_data = json.loads(json_str)
        if not isinstance(json_data, list):
            # Convert to a list to allow for common wrapping code
            json_data = [json_data]

        # Wrap the data in an "event" key
        event_list = map(self.add_event, json_data)
        # Join all the data into a single message. There is no separator
        event_data = "".join(event_list)

        try:
            reply = self.session.post(self.config['splunk.url'],
                verify = False, data=event_data,
                headers = {'Authorization': 'Splunk '
                + self.config['splunk.token']})
            reply.raise_for_status()
        except requests.exceptions.RequestException as error:
            print(f'\n*** Splunk connectivity error...aborting ***\n {error}\n')
            sys.exit(1)

        if not reply.ok:
            print(f'\n*** Failure: {reply.reason} ({reply.status_code})...aborting.\n')
            sys.exit(1)

        if self.verbose:
            print('.', end='', flush=True)

        # Return the number of sessions
        return len(json_data)

    def rabbit_callback(self, channel, method, properties, body):
        '''Callback for new messages from RabbitMQ.'''

        # Send to splunk
        session_count = self.send_to_splunk(properties.content_encoding, body)
        # Acknowledge that the message has been handled
        channel.basic_ack(delivery_tag = method.delivery_tag)
        # Publish the message count to the main process for aggregation
        self.result_queue.put((self.process_num, session_count))


def get_config(config_file):
        # Grab the parameters from the config file
        config = configparser.ConfigParser()
        config_data = {}
        if len(config.read(config_file)) == 1:
            config_data['rabbitmq.user'] = config.get('rabbitmq', 'user')
            config_data['rabbitmq.password'] = config.get('rabbitmq', 'password')
            config_data['rabbitmq.server'] = config.get('rabbitmq', 'server')
            config_data['rabbitmq.amqp_port'] = config.get('rabbitmq', 'amqp_port')
            config_data['rabbitmq.http_port'] = config.get('rabbitmq', 'http_port')
            config_data['rabbitmq.vhost'] = config.get('rabbitmq', 'vhost')
            config_data['rabbitmq.exchange'] = config.get('rabbitmq', 'exchange')

            config_data['splunk.url'] = config.get('splunk','url')
            config_data['splunk.token'] = config.get('splunk','token')

            config_data['threads'] = config.getint('general', 'threads')
            config_data['verbose'] = config.getboolean('general', 'verbose')
            config_data['debug'] = config.getboolean('general', 'debug')
        else:
            print(f'\n*** Unable to open configuration file \'{config_file}\'.')
            sys.exit(1)

        return config_data

def add_vhost(config):
    '''Create the RabbitMQ virtual host required by the SA ICDx export mechanism.'''

    # curl -i -u guest:guest -H "content-type:application/json"
    #   -XPUT http://localhost:15672/api/vhosts/dx
    url = (
        f'http://'
        f'{config["rabbitmq.server"]}:'
        f'{config["rabbitmq.http_port"]}/api/vhosts/'
        f'{config["rabbitmq.vhost"]}'
    )

    # Verify RabbitMQ connectivity while configuring the virtual host.
    try:
        reply = requests.put(url, verify = False,
                        auth = HTTPBasicAuth(config['rabbitmq.user'],
                        config['rabbitmq.password']))
    except Exception as error:
        print(f'\n*** The connection was refused to RabbitMQ ({url})...aborting. Error: ${error}\n')
        sys.exit(1)

    if not reply.ok:
        print(f'\n*** Failed to add the RabbitMQ virtual host ({reply.text})...aborting.\n')
        sys.exit(1)

def test_splunk_connection(config):
    #Debug: Uncomment the line below to enable very verbose HTTP debugging
    #add_stderr_logger()

    # Disable the warning about connecting without verifying SSL
    urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

    url=config['splunk.url']
    headers = {"Authorization": "Splunk " + config['splunk.token']}
    json_health_check = {"event": "Splunk connectivity check"}
    try:
        reply = requests.post(url, headers=headers, json=json_health_check, verify=False)
        reply.raise_for_status()
    except requests.exceptions.HTTPError as error:
        print('\n*** An HTTP error occurred while connecting to Splunk.')
        print('*** Verify the config file\'s Splunk URL path and token\n')
        print(f'{error}\n')
        sys.exit(1)
    except requests.exceptions.ConnectionError as error:
        print('\n*** An HTTP Connection error occurred while connecting to Splunk.')
        print('*** Verify the config file\'s Splunk URL IP address and port.\n')
        print(f'{error}\n')
        sys.exit(1)
    except requests.exceptions.RequestException as error:
        print('\n*** An exception occurred while connecting to Splunk.')
        print(f'{error}\n')
        sys.exit(1)

def new_process(config, process_num, result_queue):
    '''Parse the configuration file and start RabbitToSplunk() forwarding.'''
    # Python doesn't support true multi-threading due to the GIL. The way to
    # work around this limitation is to create multiple processes. This method
    # will create a new process that will run independently of the main
    # process, other than sending back results for reporting.
    try:
        # Begin forwarding
        forwarder = RabbitToSplunk(config, process_num, result_queue)
        forwarder.start()
    except KeyboardInterrupt:
        print('Interrupted. Exiting child python process.', file=sys.stderr)


def main():
    '''Get the user configuration and spawn the requested child processes.'''

    # Disable the warning about connecting without verifying SSL
    urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

    # Locate the .ini config file
    parser = argparse.ArgumentParser()
    parser.add_argument("--config", help="The configuration file", required=True)
    args = parser.parse_args()

    # Grab the parameters from the config file
    config  = get_config(args.config)

    # Configure the RabbitMQ virtual host
    add_vhost(config)

    # Make sure we can connect to splunk
    test_splunk_connection(config)

    # Create the requested number of processes
    result_queue = mp.Queue()
    process_list = map(lambda process_num:
                       mp.Process(target=new_process,
                       args=(config, process_num, result_queue)),
                       range(config['threads']))
    try:
        print('Waiting for messages. To exit press CTRL+C', file=sys.stderr)
        # Start the processes
        for p in process_list:
            p.start()

        # Counters for periodic status
        total_msg_count = 0
        last_total_msg_count = 0
        total_session_count = 0
        last_total_session_count = 0
        last_time = int(time.time())

        while True:
            # Wait for a result from one of the child processes
            (process_num, session_count) = result_queue.get()
            total_msg_count += 1
            total_session_count += session_count

            now = int(time.time())
            delta_time = now - last_time
            if delta_time >= 5:
                msg_per_second = (total_msg_count - last_total_msg_count) / delta_time
                sessions_per_second = (total_session_count - last_total_session_count) / delta_time
                print(f'\n[python] Exchange: {config["rabbitmq.exchange"]}, Total Messages: {total_msg_count}, '
                      f'Messages/Second: {msg_per_second:,.1f}, '
                      f'Sessions/Second: {int(sessions_per_second)}')
                last_time = now
                last_total_msg_count = total_msg_count
                last_total_session_count = total_session_count
    except KeyboardInterrupt:
        print('Interrupted', file=sys.stderr)
        for p in process_list:
            p.kill()
            p.join()
        sys.exit(1)


if __name__ == '__main__':
    '''Run while listenting for a CTRL+C.'''

    main()
